src/lib/HasuraGraphiQL.tsx

```diff
@@ -0,0 +1,563 @@
+  customToolbar?: React.ReactNode;
+}) {
+  const [loading, setLoading] = React.useState(true);
+  const [schema, setSchema] = React.useState<GraphQLSchema | undefined>(
+    undefined
+  );
+  const [url, setUrl] = React.useState(defaultUrl);
+  const [urlInput, setUrlInput] = React.useState(defaultUrl);
+  const [query, setQuery] = React.useState<string | undefined>(defaultQuery);
+  const [headers, setHeaders] = React.useState(
+    untransformHeaders(defaultHeaders, hiddenHeaders)
+  );
+  const [headersInput, setHeadersInput] = React.useState(
+    untransformHeaders(defaultHeaders, hiddenHeaders)
+  );
+  const [urlCollapsed, setUrlCollapsed] = React.useState(false);
+  const [headersCollapsed, setHeadersCollapsed] = React.useState(false);
+  const [relay, setRelay] = React.useState(false);
+  const [codeExporterVisible, setCodeExporterVisible] = React.useState(false);
+  const [explorerVisible, setExplorerVisible] = React.useState(true);
+  const [errorShown, setErrorShown] = React.useState(false);
+  const [error, setError] = React.useState(null);
+  const [responseTime, setResponseTime] = React.useState<number | null>(null);
+  const [responseSize, setResponseSize] = React.useState(0);
+  const [explorerWidth, setExplorerWidth] = React.useState(300);
+  const [resizing, setResizing] = React.useState(false);
+  const [isCached, setIsCached] = React.useState(false);
+
+  const updateHeaders = () => {
+    if (headersInput !== headers) {
+      setSchema(undefined);
+      setLoading(true);
+      setHeaders(headersInput);
+    }
+  };
+
+  const graphQLFetcher = makeFetcher(
+    url,
+    url.replace("http", "ws"),
+    transformHeaders(headers),
+    async function customFetch(...args) {
+      setIsCached(hasCacheDirective(query));
+      setResponseTime(null);
+      let start = Date.now();
+      let returnedPromise = fetch(...args);
+      let res = await returnedPromise;
+      setResponseTime(Date.now() - start);
+      let cloned = res.clone();
+      setResponseSize(JSON.stringify(await cloned.json()).length * 2);
+      return returnedPromise;
+    },
+    () => {
+      setResponseTime(null);
+      return null;
+    }
+  );
+
+  function ErrorNotification() {
+    return (
+      <div className="hasura-graphiql-notifications-wrapper">
+        <div className="hasura-graphiql-notification-tr">
+          <div className="hasura-graphiql-notification-inner">
+            <h4 className="hasura-graphiql-notification-title">
+              Schema Introspection Error
+            </h4>
+            <div className="hasura-graphiql-notification-message">
+              {error ? error : "Error loading schema"}
+            </div>
+            <span
+              className="hasura-graphiql-notification-dismiss"
+              onClick={() => setErrorShown(false)}
+            >
+              ×
+            </span>
+          </div>
+        </div>
+      </div>
+    );
+  }
+
+  const extraButtons = () => {
```

Comment:
I think I'll add a `toolbar:React.ReactNode` prop and use that instead of the default toolbar if it's present. Would that work?

Reply:
@beaussan it isn't. But I was thinking of this as a component; console is just an example.
@kikkupico why don't we keep the buttons and toolbars API same as what the native GraphiQL does? I don't see a reason to move away from it.

Reply:
Ok. One concern here - right now, to customise the toolbar you need to add children like GraphiQL.Toolbar and GraphiQL.Button. So, should I re-export these components in the HasuraGraphiQL lib so that the consumers can use them without having to install graphiql ?

Reply:
@kikkupico I think we should yes to make it easier to use